flink-architecture-tests/flink-architecture-tests-production/src/main/java/org/apache/flink/architecture/rules/ConnectorRules.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.architecture.rules;
+
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.thirdparty.com.google.common.base.Joiner;
+
+import static com.tngtech.archunit.base.DescribedPredicate.not;
+import static com.tngtech.archunit.core.domain.JavaModifier.PUBLIC;
+import static com.tngtech.archunit.core.domain.properties.HasModifiers.Predicates.modifier;
+import static com.tngtech.archunit.library.freeze.FreezingArchRule.freeze;
+import static org.apache.flink.architecture.common.GivenJavaClasses.noJavaClassesThat;
+import static org.apache.flink.architecture.common.SourcePredicates.areProductionCode;
+
+/** Rules for Flink connectors. */
+public class ConnectorRules {
+    private static final String[] CONNECTOR_PACKAGES = {
+        "org.apache.flink.connector..", "org.apache.flink.streaming.connectors.."
+    };
+
+    private static DescribedPredicate<JavaClass> areNotPublicAndResideOutsideOfPackages(
+            String... packageIdentifiers) {
+        return JavaClass.Predicates.resideOutsideOfPackages(packageIdentifiers)
+                .and(not(modifier(PUBLIC)))
```

Comment:
Is publicEvolving also allowed?

Reply:
The JavaDoc for PublicEvolving states
``` 
Classes and methods with this annotation are intended for public use and have stable behavior.
However, their interfaces and signatures are not considered to be stable and might be changed
across versions.```

So I think connector developers should be able to rely on them. For example the new `Sink` interface is still PublicEvolving, so if we prohibit this then no connector could depend on it for at least another release cycle

Reply:
then we should include publicEvolving here.

Reply:
It is included already because we're using the class modifier here, not the Flink annotation

Reply:
Now I got your point. we need to make clear what is the definition of "public API" for this rule:

1. class with the public visibility.
or
2. class with the public visibility and has Flink @Public or @PublicEvolving annotation.

Reply:
It should be number 2

Reply:
I updated the rule to require the Flink annotations as well